lib/rspec/core/notifications.rb

```diff
@@ -515,6 +515,54 @@ def fully_formatted(colorizer=::RSpec::Core::Formatters::ConsoleCodes)
 
         formatted
       end
+
+    private
+
+      def rerun_argument_for(example)
+        location = example.location_rerun_argument
+        return location unless duplicate_rerun_locations.include?(location)
+        conditionally_quote(example.id)
+      end
+
+      def duplicate_rerun_locations
+        @duplicate_rerun_locations ||= begin
+          locations = RSpec.world.all_examples.map(&:location_rerun_argument)
+
+          Set.new.tap do |s|
+            locations.group_by { |l| l }.each do |l, ls|
+              s << l if ls.count > 1
+            end
+          end
+        end
+      end
+
+      # Known shells that require quoting: zsh, csh, tcsh.
+      #
+      # Feel free to add other shells to this list that are known to
+      # allow `rspec ./some_spec.rb[1:1]` syntax without quoting the id.
+      #
+      # @private
+      SHELLS_ALLOWING_UNQUOTED_IDS = %w[ bash ksh fish ]
+
+      def conditionally_quote(id)
+        return id if shell_allows_unquoted_ids?
+        "'#{id.gsub("'", "\\\\'")}'"
+      end
+
+      def shell_allows_unquoted_ids?
+        return @shell_allows_unquoted_ids if defined?(@shell_allows_unquoted_ids)
+
+        @shell_allows_unquoted_ids = SHELLS_ALLOWING_UNQUOTED_IDS.include?(
+          # Note: ENV['SHELL'] isn't necessarily the shell the user is currently running.
+          # According to http://pubs.opengroup.org/onlinepubs/9699919799/basedefs/V1_chap08.html:
+          # "This variable shall represent a pathname of the user's preferred command language interpreter."
+          #
+          # It's the best we can easily do, though. We err on the side of safety (quoting
+          # the id when not actually needed) so it's not a big deal if the user is actually
+          # using a different shell.
+          ENV['SHELL'].to_s.split('/').last
```

Comment:
Yeah, it will work. I checked it, and fish behaves correctly either way, so it probably doesn't actually affect me. When I said it, I was thinking it'd be nice to say `--shell /usr/local/bin/fish`, but I guess it's more of a question of whether anything else is likely to use this variable. If not, then specifying quote preference is going to be more correct and more obvious as to what its purpose is.